src/modules/rlm_magic/rlm_magic.c

```diff
@@ -0,0 +1,249 @@
+
+/*
+ *	Define a structure for our module configuration.
+ *
+ *	These variables do not need to be in a structure, but it's
+ *	a lot cleaner to do so, and a pointer to the structure can
+ *	be used as the instance handle.
+ */
+typedef struct
+{
+	unsigned short port;
+	char const *socket_address;
+	bool is_local;
+} rlm_magic_t;
+
+/*
+ *	A mapping of configuration file names to internal variables.
+ */
+static const CONF_PARSER module_config[] = {
+	{"port", FR_CONF_OFFSET(PW_TYPE_SHORT, rlm_magic_t, port), "12345"},
+	{"is_local", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_magic_t, is_local), false},
+	{"socket-address", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_magic_t, socket_address), "127.0.0.1"},
+	CONF_PARSER_TERMINATOR};
+
+/*
+ *	Do any per-module initialization that is separate to each
+ *	configured instance of the module.  e.g. set up connections
+ *	to external databases, read configuration files, set up
+ *	dictionary entries, etc.
+ */
+static int mod_instantiate(UNUSED CONF_SECTION *conf, void *instance)
+{
+	rlm_magic_t *inst = instance;
+	DEBUG2("Creating Magic module");
+
+	if (inst->is_local)
+	{
+		#ifndef HAVE_SYS_UN_H
+			ERROR("Trying to use local unix sockets on a system that does not support it");
+			return -1;
+		#endif
+		DEBUG2("Using local Unix socket %s for authentication", inst->socket_address);
+	}
+	else
+	{
+		DEBUG2("Using remote server %s:%i for authentication", inst->socket_address, inst->port);
+	}
+
+	return 0;
+}
+
+/*
+ *	Don't currently do anything here
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authorize(UNUSED void *instance, UNUSED REQUEST *request)
+{
+	return RLM_MODULE_NOOP;
+}
+
+/*
+ *	Authenticate the user with the given password.
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(void *instance, REQUEST *request)
+{
+	// The instance variables
+	rlm_magic_t *inst = instance;
+
+	//Socket related vars
+	char buffer[BUFSIZ];
+	int sockfd;
+
+	DEBUG2("Attempting Magic Authentication");
+
+	if (inst->is_local)
+	{
+		sockfd = fr_socket_client_unix(inst->socket_address, false);
+		if (sockfd == -1)
+		{
+			fr_perror("socket");
+			return RLM_MODULE_FAIL;
+		}
+	}
+	else
+	{
+		fr_ipaddr_t address_info;
+		address_info.af = AF_INET;
+		struct hostent *hostent;
+		
+		hostent = gethostbyname(inst->socket_address);
+		if (hostent == NULL)
+		{
+			fprintf(stderr, "error: gethostbyname(\"%s\")\n", inst->socket_address);
+			return RLM_MODULE_FAIL;
+		}
+		in_addr_t in_addr = inet_addr(inet_ntoa(*(struct in_addr *)*(hostent->h_addr_list)));
+		if (in_addr == (in_addr_t)-1)
+		{
+			fprintf(stderr, "error: inet_addr(\"%s\")\n", *(hostent->h_addr_list));
+			return RLM_MODULE_FAIL;
+		}
+		address_info.ipaddr.ip4addr.s_addr = in_addr;
+
+		sockfd = fr_socket_client_tcp(NULL, &address_info, inst->port, false);
+	}
+	// Create our message which needs to be in this format
+	// '{"address": "username", "password": "password", "sessionId": ""}\n'
+	char *socket_message = talloc_array(request, char, BUFSIZ);
```

Comment:
There's no need to do multiple `strlcat`.  Just one massive `talloc_asprintf(request, " ...", args);` will do.

It's easier to understand, and less code.